### PR TITLE
536 eg comparison fmri update

### PR DESCRIPTION
**Pull request by egolol**

### CODE REVIEW

File: braph2/workflows/fMRI/test_ComparisonfMRI.m [hunk 1]
```diff
@@ -166,61 +287,187 @@
...
-            isequal(size(comparison_confidence_interval_max), [atlas.getBrainRegions().length(), atlas.getBrainRegions().length()]), ...
-            [BRAPH2.STR ':ComparisonfMRI:' BRAPH2.BUG_FUNC], ...
-            ['ComparisonfMRI does not initialize correctly with binodal measures.']) %#ok<*NBRAK>
+            ['ComparisonfMRI does not initialize correctly with binodal measures.'])
+        
+        assert(iscell(comparison_average_value_1) && ...
+            isequal(numel(comparison_average_value_1), 1) && ...
+            all(cellfun(@(x) isequal(size(x), [atlas.getBrainRegions().length(), atlas.getBrainRegions().length()]), comparison_average_value_1)) , ...
+            [BRAPH2.STR ':ComparisonMRI:' BRAPH2.BUG_FUNC], ...
+            'ComparisonMRI does not initialize correctly with binodal measures')
```
Comment (annacanal) [inline braph2/workflows/fMRI/test_ComparisonfMRI.m:429]: Comparison**f**MRI

File: braph2/workflows/fMRI/ComparisonfMRI.m [hunk 2]
```diff
@@ -99,218 +101,290 @@ function initialize_data(c, varargin)
...
                 c.average_value_1 = get_from_varargin( ...
-                    zeros(atlas.getBrainRegions().length(), 1), ...
+                    {zeros(atlas.getBrainRegions().length(), 1)}, ...
                     'ComparisonfMRI.average_value_1', ...
                     varargin{:});
+                assert(iscell(c.getGroupAverageValue(1)) && ...
+                    isequal(size(c.getGroupAverageValue(1)), [1, 1]) && ...
+                    all(cellfun(@(x) isequal(size(x), [atlas.getBrainRegions().length(), 1]), c.getGroupAverageValue(1))), ...
+                    [BRAPH2.STR ':ComparisonDTI:' BRAPH2. WRONG_INPUT], ...
+                    'Data not compatible with ComparisonDTI')
```
Comment (annacanal) [inline braph2/workflows/fMRI/ComparisonfMRI.m:213]: **ComparisonfMRI**

File: braph2/workflows/fMRI/ComparisonfMRI.m [hunk 3]
```diff
@@ -99,218 +101,290 @@ function initialize_data(c, varargin)
...
-                    ['BRAPH:ComparisonfMRI:WrongData'], ...
-                    ['Data not compatible with ComparisonfMRI.'])
-                assert(isequal(size(c.getGroupAverageValue(2)), [atlas.getBrainRegions().length(), 1]), ...
-                    ['BRAPH:ComparisonfMRI:WrongData'], ...
-                    ['Data not compatible with ComparisonfMRI.'])
+                assert(iscell(c.getGroupAverageValue(2)) && ...
+                    isequal(size(c.getGroupAverageValue(2)), [1, 1]) && ...
+                    all(cellfun(@(x) isequal(size(x), [atlas.getBrainRegions().length(), 1]), c.getGroupAverageValue(2))), ...
+                    [BRAPH2.STR ':ComparisonDTI:' BRAPH2. WRONG_INPUT], ...
+                    'Data not compatible with ComparisonDTI')
```
Comment (annacanal) [inline braph2/workflows/fMRI/ComparisonfMRI.m:223]: **ComparisonfMRI**

File: braph2/workflows/fMRI/ComparisonfMRI.m [hunk 4]
```diff
@@ -99,218 +101,290 @@ function initialize_data(c, varargin)
...
                 c.average_value_1 = get_from_varargin( ...
-                    zeros(atlas.getBrainRegions().length()), ...
+                    {zeros(atlas.getBrainRegions().length())}, ...
                     'ComparisonfMRI.average_value_1', ...
                     varargin{:});
+                assert(iscell(c.getGroupAverageValue(1)) && ...
+                    isequal(size(c.getGroupAverageValue(1)), [1, 1]) && ...
+                    all(cellfun(@(x) isequal(size(x), [atlas.getBrainRegions().length(), atlas.getBrainRegions().length()]), c.getGroupAverageValue(1))), ...
+                    [BRAPH2.STR ':ComparisonDTI:' BRAPH2. WRONG_INPUT], ...
+                    'Data not compatible with ComparisonDTI')
```
Comment (annacanal) [inline braph2/workflows/fMRI/ComparisonfMRI.m:315]: **ComparisonfMRI**

File: braph2/workflows/fMRI/ComparisonfMRI.m [hunk 5]
```diff
@@ -99,218 +101,290 @@ function initialize_data(c, varargin)
...
-                    ['BRAPH:ComparisonfMRI:WrongData'], ...
-                    ['Data not compatible with ComparisonfMRI.'])
-                assert(isequal(size(c.getGroupAverageValue(2)), [atlas.getBrainRegions().length(), atlas.getBrainRegions().length()]), ...
-                    ['BRAPH:ComparisonfMRI:WrongData'], ...
-                    ['Data not compatible with ComparisonfMRI.'])
+                assert(iscell(c.getGroupAverageValue(2)) && ...
+                    isequal(size(c.getGroupAverageValue(2)), [1, 1]) && ...
+                    all(cellfun(@(x) isequal(size(x), [atlas.getBrainRegions().length(), atlas.getBrainRegions().length()]), c.getGroupAverageValue(2))), ...
+                    [BRAPH2.STR ':ComparisonDTI:' BRAPH2. WRONG_INPUT], ...
+                    'Data not compatible with ComparisonDTI')
```
Comment (annacanal) [inline braph2/workflows/fMRI/ComparisonfMRI.m:325]: **ComparisonfMRI**

File: braph2/workflows/fMRI/ComparisonfMRI.m [hunk 6]
```diff
@@ -99,218 +101,290 @@ function initialize_data(c, varargin)
...
                 c.all_differences = get_from_varargin( ...
                     repmat({zeros(atlas.getBrainRegions().length())}, 1, number_of_permutations), ...
                     'ComparisonfMRI.all_differences', ...
                     varargin{:});
+                assert(iscell(c.getAllDifferences()) && ...
+                    isequal(size(c.getAllDifferences()), [1, number_of_permutations]) && ...
+                    all(cellfun(@(x) isequal(size(x), [atlas.getBrainRegions().length(), atlas.getBrainRegions().length()]), c.getAllDifferences())), ...
+                    [BRAPH2.STR ':ComparisonDTI:' BRAPH2. WRONG_INPUT], ...
+                    'Data not compatible with ComparisonDTI')
+
```
Comment (annacanal) [inline braph2/workflows/fMRI/ComparisonfMRI.m:348]: **ComparisonfMRI**

File: braph2/workflows/fMRI/ComparisonfMRI.m [hunk 7]
```diff
@@ -99,218 +101,290 @@ function initialize_data(c, varargin)
...
                 c.difference = get_from_varargin( ...
-                    zeros(atlas.getBrainRegions().length()), ...
+                    {zeros(atlas.getBrainRegions().length())}, ...
                     'ComparisonfMRI.difference', ...
                     varargin{:});
+                assert(iscell(c.getDifference()) && ...
+                    isequal(size(c.getDifference()), [1, 1]) && ...
+                    all(cellfun(@(x) isequal(size(x), [atlas.getBrainRegions().length(), atlas.getBrainRegions().length()]), c.getDifference())), ...
+                    [BRAPH2.STR ':ComparisonDTI:' BRAPH2. WRONG_INPUT], ...
+                    'Data not compatible with ComparisonDTI')
```
Comment (annacanal) [inline braph2/workflows/fMRI/ComparisonfMRI.m:337]: **ComparisonfMRI**

File: braph2/workflows/fMRI/ComparisonfMRI.m [hunk 8]
```diff
@@ -99,218 +101,290 @@ function initialize_data(c, varargin)
...
                 c.p1 = get_from_varargin( ...
-                    zeros(atlas.getBrainRegions().length()), ...
+                    {zeros(atlas.getBrainRegions().length())}, ...
                     'ComparisonfMRI.p1', ...
                     varargin{:});
+                assert(iscell(c.getP1()) && ...
+                    isequal(size(c.getP1()), [1, 1]) && ...
+                    all(cellfun(@(x) isequal(size(x), [atlas.getBrainRegions().length(), atlas.getBrainRegions().length()]), c.getP1())), ...
+                    [BRAPH2.STR ':ComparisonDTI:' BRAPH2. WRONG_INPUT], ...
+                    'Data not compatible with ComparisonDTI')
```
Comment (annacanal) [inline braph2/workflows/fMRI/ComparisonfMRI.m:357]: **ComparisonfMRI**

File: braph2/workflows/fMRI/ComparisonfMRI.m [hunk 9]
```diff
@@ -99,218 +101,290 @@ function initialize_data(c, varargin)
...
                 c.p2 = get_from_varargin( ...
-                    zeros(atlas.getBrainRegions().length()), ...
+                    {zeros(atlas.getBrainRegions().length())}, ...
                     'ComparisonfMRI.p2', ...
                     varargin{:});
+                assert(iscell(c.getP2()) && ...
+                    isequal(size(c.getP2()), [1, 1]) && ...
+                    all(cellfun(@(x) isequal(size(x), [atlas.getBrainRegions().length(), atlas.getBrainRegions().length()]), c.getP2())), ...
+                    [BRAPH2.STR ':ComparisonDTI:' BRAPH2. WRONG_INPUT], ...
+                    'Data not compatible with ComparisonDTI')
```
Comment (annacanal) [inline braph2/workflows/fMRI/ComparisonfMRI.m:367]: **ComparisonfMRI**

File: braph2/workflows/fMRI/AnalysisfMRI.m [hunk 10]
```diff
@@ -205,112 +215,77 @@
...
+            all_permutations_1 = cell(1, M);
+            all_permutations_2 = cell(1, M);
+            
+            start = tic;
+            for i = 1:1:M
+                if verbose
+                    disp(['** PERMUTATION TEST - sampling #' int2str(i) '/' int2str(M) ' - ' int2str(toc(start)) '.' int2str(mod(toc(start),1)*10) 's'])
+                end
+                
+                if is_longitudinal
```
Comment (giovannivolpe) [inline braph2/workflows/fMRI/AnalysisfMRI.m:244]: There is not Permutation any more. This should give an error when tested.

File: braph2/workflows/fMRI/AnalysisfMRI.m [hunk 11]
```diff
@@ -205,112 +215,77 @@
...
+                end
+            end
+            
+            difference_all_permutations = cellfun(@(x, y) y - x, all_permutations_1, all_permutations_2, 'UniformOutput', false);
+            p1 = {pvalue1(difference_mean{1}, difference_all_permutations)};  % singe tail,
+            p2 = {pvalue2(difference_mean{1}, difference_all_permutations)};  % double tail
+            qtl = quantiles(difference_all_permutations, 40);
+            ci_lower = {cellfun(@(x) x(2), qtl)};
+            ci_upper = {cellfun(@(x) x(40), qtl)};
+
```
Comment (giovannivolpe) [inline braph2/workflows/fMRI/AnalysisfMRI.m:267]: remove extra space

File: braph2/workflows/fMRI/AnalysisfMRI.m (hunk 10 above)
Comment (giovannivolpe) [reply braph2/workflows/fMRI/AnalysisfMRI.m:244]: All this block should be:
[permuted_values_1, permuted_values_2] = permutation(values1, values2, is_longitudinal);

**Pull request merged**